doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+
+Nbrmgrd will also have a peridic neighbor timer to handle both the Neighbor refresh and the FDB refresh. If a neighbor's refresh timer expires then ARP request or Neighbor solicitation message will be sent using a RAW socket. Nbrmgrd will not wait for the response because waiting for a response will delay sending of subsequent refresh packets and in scale scenario will impact refresh process. Once the packet is transmitted out successfully the response from the destination host will be processed by kernel and kernel will update its internal timers.
+
+## 3.2 DB Changes
+
+### 3.2.1 CONFIG DB
+Not applicable
+
+### 3.2.2 APP DB
+**NEIGH_REFRESH_TABLE**
+
+```
+Producer: neighsyncd 
+
+Consumer: nbrmgrd
+
+Description: Create a neighbor refresh table entry for every neighbor that is added to NEIGH_TABLE 
+```
+
+Schema:
+
+```
+; New table
+; Has the ARP/ND entries that needs to be refreshed periodically
+
+key           = NEIGH_REFRESH_TABLE:ifname:IPprefix
+
+; field       = value
+neigh         = 12HEXDIG           ; mac address of the neighbor
+family        = "IPv4" / "IPv6"    ; address family
+state         = 1*10DIGIT          ; Neighbor state in Kernel
+flags         = 1*10DIGIT          ; Neighbor flag info in Kernel
+
+Example:
+127.0.0.1:63792> hgetall "NEIGH_REFRESH_TABLE:Ethernet25:20.20.20.1"
+ 1) "neigh"
+ 2) "80:a2:35:26:0c:61"
+ 3) "family"
+ 4) "IPv4"
+ 5) "state"
+ 6) "2"
+ 7) "flags"
+ 8) "0"
+127.0.0.1:63792>
+```
+
+### 3.2.3 STATE DB
+Not applicable
+
+### 3.2.4 ASIC DB
+Not applicable
+
+### 3.2.5 COUNTER DB
+Not applicable
+
+### 3.2.6 Configuration save and restart
+Not applicable
+
+## 3.3 Switch State Service Design
+
+### 3.3.1 Orchestration Agent
+
+No Change in Orchangent / NeighOrch.
+
+### 3.3.2 NeighSyncd
+
+Neighsyncd is responsible for syncing the kernel neighbor table to the hardware via the APP_DB and OrchAgents. Neighsyncd listens on netlink events (RTM_NEWNEIGH, RTM_DELNEIGH) and creates/deletes NEIGH_TABLE entries in APP_DB.
+
+Existing functionality of neighsyncd is retained as it is. In addition to managing NEIGH_TABLE entries in APP_DB, neighsyncd will also add the details of the neighbor into NEIGH_REFRESH_TABLE.
+
+
+### 3.3.2 Nbrmgrd
+
+Nbrmgrd in addition to its existing responsibilities will now have to take care of the following Neighbor Refresh resonsibilities.
+
+- listens to NEIGH_REFRESH_TABLE and manage a local cache of neighbor entries
+- subscribe to redis-db interface CONFIG_DB tables to gather the data required to send the ARP refresh packets
+- subscribe to STATE_DB FDB table to manage the mac address table local cache
+- periodically refresh ARP/ND by sending ARP request pkt / NS pkt
+
+
+#### 3.3.2.1 Local Cache
+
+Neighbor Refresh Cache:
+
+- Add neighbor entries to cache when the entry is added to NEIGH_REFRESH_TABLE
+    - All Dynamically learned neighbor entries [ARP, ND (Global, LinkLocal)]
+    - All Static neighbor entries (MAC can be dynamic)
```

Comment:
I wonder why do we need to support this case, what if we get different than configured MAC from the peer as part of the refresh? cant we mandate MAC also be static, any use-case?

Reply:
This is a special case and not a regular use-case scenario.

Static Neigh and Static MAC are independent, hence it is possible to have one without the other.

Nbr refresh module will look for the MAC in the static neighbor config, hence even if the ARP response is received with different MAC (misconfig), kernel will not accept the pkt as the neighbor is PERMANENT and nbr refresh module will retry until the MAC given in the static neighbor config is learned.